src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -288,112 +319,46 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
       }
       // Otherwise, leave result == nullptr because the adjusted size is smaller than min size.
     } else {
       // This is a GCLAB or a TLAB allocation
+      size_t size = req.size();
       size_t free = align_down(r->free() >> LogHeapWordSize, MinObjAlignment);
       if (size > free) {
         size = free;
       }
       if (size >= req.min_size()) {
         result = r->allocate(size, req);
-        if (result != nullptr) {
-          // Record actual allocation size
-          req.set_actual_size(size);
-        }
         assert (result != nullptr, "Allocation must succeed: free " SIZE_FORMAT ", actual " SIZE_FORMAT, free, size);
+        req.set_actual_size(size);
       } else {
         log_trace(gc, free)("Failed to shrink TLAB or GCLAB request (" SIZE_FORMAT ") in region " SIZE_FORMAT " to " SIZE_FORMAT
                            " because min_size() is " SIZE_FORMAT, req.size(), r->index(), size, req.min_size());
       }
     }
   } else if (req.is_lab_alloc() && req.type() == ShenandoahAllocRequest::_alloc_plab) {
-    assert(_heap->mode()->is_generational(), "PLABs are only for generational mode");
+
     // inelastic PLAB
-    size_t free = r->free();
-    size_t usable_free = (free / CardTable::card_size()) << CardTable::card_shift();
-    free /= HeapWordSize;
-    usable_free /= HeapWordSize;
-    if ((free != usable_free) && (free - usable_free < ShenandoahHeap::min_fill_size() * HeapWordSize)) {
-      // We'll have to add another card's memory to the padding
-      if (usable_free > CardTable::card_size_in_words()) {
-        usable_free -= CardTable::card_size_in_words();
-      } else {
-        assert(usable_free == 0, "usable_free is a multiple of card_size and card_size > min_fill_size");
-      }
-    }
-    assert(size % CardTable::card_size_in_words() == 0, "PLAB size must be multiple of remembered set card size");
+    size_t size = req.size();
+    size_t usable_free = get_usable_free_words(r->free());
     if (size <= usable_free) {
-      result = r->allocate_aligned(size, req, CardTable::card_size());
-      size = req.actual_size();
-      assert(result != nullptr, "Allocation cannot fail");
-      assert(r->top() <= r->end(), "Allocation cannot span end of region");
-      assert(req.actual_size() % CardTable::card_size_in_words() == 0, "PLAB start must align with card boundary");
-      assert(((uintptr_t) result) % CardTable::card_size_in_words() == 0, "PLAB start must align with card boundary");
-      if (free > usable_free) {
-        // Account for the alignment padding
-        size_t padding = (free - usable_free) * HeapWordSize;
-        increase_used(padding);
-        assert(r->is_old(), "All PLABs reside in old-gen");
-        _heap->old_generation()->increase_used(padding);
-        // For verification consistency, we need to report this padding to _heap
-        _heap->increase_used(padding);
-      }
+      result = allocated_aligned_plab(size, req, r);
```

Comment:
`allocate_..`

Reply:
Isn't this likely to fail -- at least in the general case -- the alignment checks in assertions in `allocate_aligned_plab()`?

I am not sure I understand the difference between `ShenandoahElasticPLAB` and not, since the documentation comment isn't very helpful:
```
  product(bool, ShenandoahElasticTLAB, true, DIAGNOSTIC,                    \
          "Use Elastic TLABs with Shenandoah")                              \
```

Reply:
When `ShenandoahElasticTLAB` is enabled, the allocator is allowed to return a TLAB smaller than the requested size (but greater than the `min_size`).